Userland/Libraries/LibGUI/FileSystemModel.cpp

```diff
@@ -668,43 +671,65 @@ bool FileSystemModel::fetch_thumbnail_for(Node const& node)
         return true;
     }
 
-    // Otherwise, arrange to render the thumbnail
-    // in background and make it available later.
+    // See if the thumbnail was loaded, but the callback was never invoked.
+    // This often happens if a window with a FileSystemModel is closed and later reopened.
+    auto promise_it = s_loading_thumbnails.find(path);
+    if (promise_it != s_loading_thumbnails.end()) {
+        auto const& result = promise_it->value->result();
+        // The promise was successfully resolved since we started it.
+        if (result.has_value()) {
+            s_loading_thumbnails.remove(path);
+            s_thumbnail_cache.set(path, result.value());
+            node.thumbnail = s_thumbnail_cache.get(path).value();
+            return true;
+        }
+        // The promise was cancelled, load this thumbnail again.
+        if (promise_it->value->is_cancelled()) {
+            s_loading_thumbnails.remove(path);
+        } else {
+            return false;
+        }
+    }
+
+    // Otherwise, arrange to render the thumbnail in background and make it available later.
 
-    s_thumbnail_cache.set(path, nullptr);
     m_thumbnail_progress_total++;
 
     auto weak_this = make_weak_ptr();
 
-    (void)Threading::BackgroundAction<ErrorOr<NonnullRefPtr<Gfx::Bitmap>>>::construct(
-        [path](auto&) {
-            return render_thumbnail(path);
-        },
-
-        [this, path, weak_this](auto thumbnail_or_error) -> ErrorOr<void> {
-            if (thumbnail_or_error.is_error()) {
-                s_thumbnail_cache.set(path, nullptr);
-                dbgln("Failed to load thumbnail for {}: {}", path, thumbnail_or_error.error());
-            } else {
-                s_thumbnail_cache.set(path, thumbnail_or_error.release_value());
-            }
+    s_loading_thumbnails.set(path,
+        BitmapBackgroundAction::construct(
+            [path](auto&) {
+                return render_thumbnail(path);
+            },
+
+            [this, path, weak_this](auto thumbnail) -> ErrorOr<void> {
+                s_loading_thumbnails.remove(path);
+                s_thumbnail_cache.set(path, thumbnail);
+
+                // The model was destroyed, no need to update
+                // progress or call any event handlers.
+                if (weak_this.is_null())
+                    return {};
```

Comment:
The way we normally do this pattern is 
```
    m_websocket->on_open = [weak_this = make_weak_ptr()] {
        if (auto strong_this = weak_this.strong_ref())
            if (strong_this->on_open)
                strong_this->on_open();
    };
```
instead of capturing both weak_this and this.